internal/storage/streamed_chunk_uploader.go

```diff
@@ -0,0 +1,206 @@
+	wc := obj.NewWriter(ctx)
+	// NewWriter never returns nil, so no nil-check here.
+
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		sow.totalWriteSucceededSoFar = n
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if sow.userProgressFunc != nil {
+			sow.userProgressFunc(n)
+		}
+	}
+	sow.wc = wc
+	sow.state = Initialized
+
+	return &sow, nil
+}
+
+func (sow StreamedChunkUploader) ChunkSize() (int, error) {
+	if nil == sow.wc {
+		return 0, fmt.Errorf("writer not initialized")
+	}
+	return sow.wc.ChunkSize, nil
+}
+
+func (sow StreamedChunkUploader) State() UploaderState {
+	return sow.state
+}
+
+func (sow StreamedChunkUploader) BytesUploadInitiatedSoFar() int64 {
+	return sow.totalWriteInitiatedSoFar
+}
+
+func (sow StreamedChunkUploader) BytesUploadedSoFar() int64 {
+	return sow.totalWriteSucceededSoFar
+}
+
+func (sow *StreamedChunkUploader) readyToWrite() bool {
+	switch sow.state {
+	case Initialized:
+		fallthrough
+	case Waiting:
+		fallthrough
+	case Writing:
+		return true
+	default:
+		return false
+	}
+}
+
+func (sow *StreamedChunkUploader) readyToClose() bool {
+	switch sow.state {
+	case Initialized:
+		fallthrough
+	case Waiting:
+		fallthrough
+	case Writing:
+		fallthrough
+	case WriteError:
+		return true
+	default:
+		return false
+	}
+}
+
+// Write uploads the given content
+// to underlying GCS object with the initial configures chunk-size,
+// number of bytes, an object writer, and a content-reader.
+// It is assumed that all the relevant attributes of the object
+// writer have been already by a previous calls to CreateObject
+// which created the object-writer.
+func (sow *StreamedChunkUploader) UploadChunkAsync(contents io.Reader) error {
+	if !sow.readyToWrite() || (sow.wc == nil) {
+		return fmt.Errorf("writer not ready to write: object: %s, status = %v, writer: %v", sow.objectName, sow.state, sow.wc)
+	}
+
+	n, err := io.Copy(sow.wc, contents)
+	sow.totalWriteInitiatedSoFar += n
+	if n == 0 || err == io.EOF {
+		// for io.EOF, don't change/append the error (message) as it makes it difficult to identify EOF for the caller.
+		sow.state = Waiting
+		return io.EOF
+	}
+	if err != nil {
+		sow.state = WriteError
+		return fmt.Errorf("totalSizeUploaded-so-far=%d, successfully-uploaded-in-last-upload=%d, chunk-size=%d for object %s: %v", sow.totalWriteSucceededSoFar, n, sow.wc.ChunkSize,
+			sow.objectName, err)
+	}
+
+	sow.state = Writing
+	return nil
+}
+
+func (sow *StreamedChunkUploader) Close() (o *gcs.Object, err error) {
+	defer func() {
+		sow.state = Destroyed
+		sow.wc = nil
+	}()
+
+	if !sow.readyToClose() {
+		return nil, fmt.Errorf("improper state for finalizing object %s", sow.objectName)
+	}
+
+	// We can't use defer to close the writer, because we need to close the
+	// writer successfully before calling Attrs() method of writer.
```

Comment:
this comment doesn't make any sense. this method is triggered to finalize the upload and it has to be done synchronously.

Reply:
Removed the unnecessary and confusing comment.